### PR TITLE
[Migration-Controller]Indicate resource quota rejection

### PR DESCRIPTION
**Pull request by Barakmor1**
- Add condition that indicates that migration was rejected by ResourceQuota
- Add resourceQuota informer within the migration controller to get fast response 
whenever resourceQuota blocked the migration and than was deleted/modified
the migration controller should respond quickly.

**What this PR does / why we need it**:
In a multi-tenant environment, a cluster administrator creates a namespace for each tenant and controls the amount of resources that they can use by creating a ResourceQuota object within the namespace.

During the migration process, there is a hidden cost involved in creating a new virt-launcher pod before terminating the old one. This differs from how standard pods are moved and can result in unexpected resource utilization from the user's perspective. This poses a challenge where the hidden cost of live migration can prevent live migrating a VM for reasons not reflected to the VM creator / namespace owner.

Migrating a virtual machine in the presence of a resource quota can cause the migration to fail and subsequently cause the upgrade to fail.

As mentioned in :
https://kubernetes.io/docs/concepts/policy/resource-quotas/#quota-and-cluster-capacity

Kubernetes recommendation for "complex policies" such as Live Migration is writing a "controller" that watches the quota usage and adjusts the quota hard limits of each namespace according to other signals.

In order to build such controller we need the migration controller to provide the following benefits:

- Ability to detect when a resourceQuota blocks a migration
- Quick response once the resourceQuota is no longer blocking a migration

Example for such controller:
https://github.com/Barakmor1/Managed-Tenant-Quota

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Add condition to migrations that indicates that migration was rejected by ResourceQuota
```

### CODE REVIEW

**Comment by Barakmor1** [comment]
@vladikr

**Comment by vladikr** [comment]
@xpivarc could you please take a look?

File: tests/migration_test.go
Comment (vladikr) [inline tests/migration_test.go:4558]: @Barakmor1 Does this have to be serial?

Comment (Barakmor1) [reply tests/migration_test.go:4558]: HEy thanks for reviewing,
I think so because the resource Quota might interfere other tests .

File: pkg/virt-controller/watch/migration.go
Comment (vladikr) [inline pkg/virt-controller/watch/migration.go:1438]: This is needed since we have an exponential backoff for failed migrations. Knowing explicitly that the failure was due to quota, which has been updated now, would speed up the migration.

**Review by fossedihelm — COMMENTED** [review]
Thanks @Barakmor1 for the improvement.
Some comments below

File: tests/migration_test.go
Comment (Barakmor1) [reply tests/migration_test.go:4558]: Actually i'm not sure, are the tests running in the same namespace when they run in parallel?
If each test get it's own namespace i guess we can remove the serial label.

Comment (xpivarc) [reply tests/migration_test.go:4558]: Yes, each test is running in its own  namespaces

File: tests/migration_test.go
Comment (Barakmor1) [reply tests/migration_test.go]: Done

File: pkg/virt-controller/watch/migration.go
Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go:186]: Your right.
Done :)

File: tests/migration_test.go
Comment (Barakmor1) [reply tests/migration_test.go]: OK than, i removed the Serial label.

**Review by fossedihelm — COMMENTED** [review]
Thanks @Barakmor1! One nit! Feel free to unhold.
/lgtm
/hold

File: tests/migration_test.go
Comment (xpivarc) [inline tests/migration_test.go:4575]: This can be simplified as follows:
1. Create VMI
2. Create a quota allowing 0 or minimum
3. Run migration and expect failure

**Review by xpivarc — CHANGES_REQUESTED** [review]
I have left a few suggestions on the tests.

@vladikr I think the 3rd commit is unnecessary. We could just re-queue when we see quota rejection

**Comment by Barakmor1** [comment]
> I have left a few suggestions on the tests.
> 
> @vladikr I think the 3rd commit is unnecessary. We could just re-queue when we see quota rejection

but in case nothing changes in the cluster we would just keep re-queuing, that is Inefficient the only thing that could unblock that migration is resourceQuota deletion or resourceQuota.status.hard\ resourceQuota.spec.hard update .

**Comment by xpivarc** [comment]
> > I have left a few suggestions on the tests.
> > @vladikr I think the 3rd commit is unnecessary. We could just re-queue when we see quota rejection
> 
> but in case nothing changes in the cluster we would just keep re-queuing and that is Inefficient , the only thing that could unblock that migration is resourceQuota deletion or resourceQuota.status.hard\ resourceQuota.spec.hard update .

As far as understand it will still happen, you did not address it. Still, we should have a timeout that will be triggered.

File: tests/migration_test.go
Comment (Barakmor1) [reply tests/migration_test.go:4604]: Maybe the Api server won't be responsive 

> > > I have left a few suggestions on the tests.
> > > @vladikr I think the 3rd commit is unnecessary. We could just re-queue when we see quota rejection
> > 
> > 
> > but in case nothing changes in the cluster we would just keep re-queuing and that is Inefficient , the only thing that could unblock that migration is resourceQuota deletion or resourceQuota.status.hard\ resourceQuota.spec.hard update .
> 
> As far as understand it will still happen, you did not address it. Still, we should have a timeout that will be triggered.

it will happen but with back-off logic that could delay the migration in case of status change

File: tests/migration_test.go
Comment (xpivarc) [reply tests/migration_test.go:4575]: Fair, you might want to fetch the pod and use the memory request then to hide the magic calculation.

File: tests/migration_test.go
Comment (Barakmor1) [reply tests/migration_test.go:4599]: I removed the timeout.
> Please don't ignore any errors. We should not program defensively here.
can you please elaborate?

File: pkg/virt-controller/watch/migration.go
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go:1440]: This can be any general update. Do you want to check if the resources were increased?

File: pkg/virt-controller/watch/migration.go
Comment (xpivarc) [inline pkg/virt-controller/watch/migration.go:1443]: These callbacks should not block for a significant time. Consider wrapping this in a goroutine.

**Review by xpivarc — COMMENTED** [review]
Please squash bdf1a1b788d825d511a6b6d4ed22abe13a1cec8e

File: pkg/virt-controller/watch/migration.go
Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go:1440]: some pods might be deleted as well this could also unblock the migration

File: pkg/virt-controller/watch/migration.go
Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go:1443]: IMHO this is unnecessary complexity.
@vladikr  WDTY?

**Comment by Barakmor1** [comment]
> Please squash https://github.com/kubevirt/kubevirt/commit/bdf1a1b788d825d511a6b6d4ed22abe13a1cec8e

I don't think that this is necessary.

**Comment by acardace** [comment]
> > Please squash [bdf1a1b](https://github.com/kubevirt/kubevirt/commit/bdf1a1b788d825d511a6b6d4ed22abe13a1cec8e)
> 
> I don't think that this is necessary.

I'm with @xpivarc  here, rebasing and squashing takes 2 seconds and enhances the readability of both the PR and the git history of the project.

**Comment by xpivarc** [comment]
> > > Please squash [bdf1a1b](https://github.com/kubevirt/kubevirt/commit/bdf1a1b788d825d511a6b6d4ed22abe13a1cec8e)
> > 
> > 
> > I don't think that this is necessary.
> 
> I'm with @xpivarc here, rebasing and squashing takes 2 seconds and enhances the readability of both the PR and the git history of the project.

Thanks, Antonio, I would just add that it is much easier to review the tests if they are within the same commit. I can see how sometimes that can enlarge the commit and it might be better to keep it separate. But it is not the case here. Apologies for not including reasoning in the first place. 

Last but not least "Adjust unit tests to the new resourceQuota informer " did not enlighten me.

Note: You don't even need to rebase, a simple reset and amend will do.

**Comment by Barakmor1** [comment]
> > > > Please squash [bdf1a1b](https://github.com/kubevirt/kubevirt/commit/bdf1a1b788d825d511a6b6d4ed22abe13a1cec8e)
> > > 
> > > 
> > > I don't think that this is necessary.
> > 
> > 
> > I'm with @xpivarc here, rebasing and squashing takes 2 seconds and enhances the readability of both the PR and the git history of the project.
> 
> Thanks, Antonio, I would just add that it is much easier to review the tests if they are within the same commit. I can see how sometimes that can enlarge the commit and it might be better to keep it separate. But it is not the case here. Apologies for not including reasoning in the first place.
> 
> Last but not least "Adjust unit tests to the new resourceQuota informer " did not enlighten me.
> 
> Note: You don't even need to rebase, a simple reset and amend will do.

Done.

File: pkg/virt-controller/watch/migration.go
Comment (Barakmor1) [reply pkg/virt-controller/watch/migration.go]: Fixed :)

File: tests/migration_test.go
Comment (xpivarc) [reply tests/migration_test.go:4640]: The `Create` call should succeed. Please remove the `Eventually` block.

**Pull request merged**
